// @tiptap/react/src/EditorContent.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var jsxRuntime = require('@emotion/react/jsx-runtime');
var React = require('react');
var ReactDOM = require('react-dom');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var ReactDOM__default = /*#__PURE__*/_interopDefaultLegacy(ReactDOM);

var Portals = function (_a) {
    var renderers = _a.renderers;
    return (jsxRuntime.jsx(jsxRuntime.Fragment, { children: Array.from(renderers).map(function (_a) {
            var _b = tslib.__read(_a, 2), key = _b[0], renderer = _b[1];
            return ReactDOM__default["default"].createPortal(renderer.reactElement, renderer.element, key);
        }) }));
};
var PureEditorContent = /** @class */ (function (_super) {
    tslib.__extends(PureEditorContent, _super);
    function PureEditorContent(props) {
        var _this = _super.call(this, props) || this;
        _this.editorContentRef = React__default["default"].createRef();
        _this.state = {
            renderers: new Map(),
        };
        return _this;
    }
    PureEditorContent.prototype.componentDidMount = function () {
        this.init();
    };
    PureEditorContent.prototype.componentDidUpdate = function () {
        this.init();
    };
    PureEditorContent.prototype.init = function () {
        var editor = this.props.editor;
        if (editor && editor.options.element) {
            if (editor.contentComponent) {
                return;
            }
            var element = this.editorContentRef.current;
            element.append.apply(element, tslib.__spreadArray([], tslib.__read(editor.options.element.childNodes), false));
            editor.setOptions({
                element: element,
            });
            editor.contentComponent = this;
            editor.createNodeViews();
        }
    };
    PureEditorContent.prototype.componentWillUnmount = function () {
        var editor = this.props.editor;
        if (!editor) {
            return;
        }
        if (!editor.isDestroyed) {
            editor.view.setProps({
                nodeViews: {},
            });
        }
        editor.contentComponent = null;
        if (!editor.options.element.firstChild) {
            return;
        }
        var newElement = document.createElement('div');
        newElement.append.apply(newElement, tslib.__spreadArray([], tslib.__read(editor.options.element.childNodes), false));
        editor.setOptions({
            element: newElement,
        });
    };
    PureEditorContent.prototype.render = function () {
        var _a = this.props; _a.editor; var rest = tslib.__rest(_a, ["editor"]);
        return (jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [jsxRuntime.jsx("div", tslib.__assign({ ref: this.editorContentRef }, rest)), jsxRuntime.jsx(Portals, { renderers: this.state.renderers })] }));
    };
    return PureEditorContent;
}(React__default["default"].Component));
var EditorContent = React__default["default"].memo(PureEditorContent);

exports.EditorContent = EditorContent;
exports.PureEditorContent = PureEditorContent;
