// @tiptap/react/src/FloatingMenu.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var jsxRuntime = require('@emotion/react/jsx-runtime');
var React = require('react');
var tiptapExtensionFloatingMenu_esm = require('../../extension-floating-menu/dist/tiptap-extension-floating-menu.esm.js');

var FloatingMenu = function (props) {
    var _a = tslib.__read(React.useState(null), 2), element = _a[0], setElement = _a[1];
    React.useEffect(function () {
        if (!element) {
            return;
        }
        if (props.editor.isDestroyed) {
            return;
        }
        var _a = props.pluginKey, pluginKey = _a === void 0 ? 'floatingMenu' : _a, editor = props.editor, _b = props.tippyOptions, tippyOptions = _b === void 0 ? {} : _b, _c = props.shouldShow, shouldShow = _c === void 0 ? null : _c;
        var plugin = tiptapExtensionFloatingMenu_esm.FloatingMenuPlugin({
            pluginKey: pluginKey,
            editor: editor,
            element: element,
            tippyOptions: tippyOptions,
            shouldShow: shouldShow,
        });
        editor.registerPlugin(plugin);
        return function () { return editor.unregisterPlugin(pluginKey); };
    }, [
        props.editor,
        element,
    ]);
    return (jsxRuntime.jsx("div", tslib.__assign({ ref: setElement, className: props.className, style: { visibility: 'hidden' } }, { children: props.children })));
};

exports.FloatingMenu = FloatingMenu;
