// swiper/shared/dom.js
'use strict';

var dom7_esm = require('../../dom7/dom7.esm.js');

const Methods = {
  addClass: dom7_esm.addClass,
  removeClass: dom7_esm.removeClass,
  hasClass: dom7_esm.hasClass,
  toggleClass: dom7_esm.toggleClass,
  attr: dom7_esm.attr,
  removeAttr: dom7_esm.removeAttr,
  transform: dom7_esm.transform,
  transition: dom7_esm.transition,
  on: dom7_esm.on,
  off: dom7_esm.off,
  trigger: dom7_esm.trigger,
  transitionEnd: dom7_esm.transitionEnd,
  outerWidth: dom7_esm.outerWidth,
  outerHeight: dom7_esm.outerHeight,
  styles: dom7_esm.styles,
  offset: dom7_esm.offset,
  css: dom7_esm.css,
  each: dom7_esm.each,
  html: dom7_esm.html,
  text: dom7_esm.text,
  is: dom7_esm.is,
  index: dom7_esm.index,
  eq: dom7_esm.eq,
  append: dom7_esm.append,
  prepend: dom7_esm.prepend,
  next: dom7_esm.next,
  nextAll: dom7_esm.nextAll,
  prev: dom7_esm.prev,
  prevAll: dom7_esm.prevAll,
  parent: dom7_esm.parent,
  parents: dom7_esm.parents,
  closest: dom7_esm.closest,
  find: dom7_esm.find,
  children: dom7_esm.children,
  filter: dom7_esm.filter,
  remove: dom7_esm.remove
};
Object.keys(Methods).forEach(methodName => {
  Object.defineProperty(dom7_esm["default"].fn, methodName, {
    value: Methods[methodName],
    writable: true
  });
});

module.exports = dom7_esm["default"];
