// swiper/shared/get-support.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var ssrWindow_esm = require('../../ssr-window/ssr-window.esm.js');

let support;

function calcSupport() {
  const window = ssrWindow_esm.getWindow();
  const document = ssrWindow_esm.getDocument();
  return {
    smoothScroll: document.documentElement && 'scrollBehavior' in document.documentElement.style,
    touch: !!('ontouchstart' in window || window.DocumentTouch && document instanceof window.DocumentTouch),
    passiveListener: function checkPassiveListener() {
      let supportsPassive = false;

      try {
        const opts = Object.defineProperty({}, 'passive', {
          // eslint-disable-next-line
          get() {
            supportsPassive = true;
          }

        });
        window.addEventListener('testPassiveListener', null, opts);
      } catch (e) {// No support
      }

      return supportsPassive;
    }(),
    gestures: function checkGestures() {
      return 'ongesturestart' in window;
    }()
  };
}

function getSupport() {
  if (!support) {
    support = calcSupport();
  }

  return support;
}

exports.getSupport = getSupport;
