// @tiptap/extension-text-style/dist/tiptap-extension-text-style.esm.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tiptapCore_esm = require('../../core/dist/tiptap-core.esm.js');

const TextStyle = tiptapCore_esm.Mark.create({
    name: 'textStyle',
    addOptions() {
        return {
            HTMLAttributes: {},
        };
    },
    parseHTML() {
        return [
            {
                tag: 'span',
                getAttrs: element => {
                    const hasStyles = element.hasAttribute('style');
                    if (!hasStyles) {
                        return false;
                    }
                    return {};
                },
            },
        ];
    },
    renderHTML({ HTMLAttributes }) {
        return ['span', tiptapCore_esm.mergeAttributes(this.options.HTMLAttributes, HTMLAttributes), 0];
    },
    addCommands() {
        return {
            removeEmptyTextStyle: () => ({ state, commands }) => {
                const attributes = tiptapCore_esm.getMarkAttributes(state, this.type);
                const hasStyles = Object.entries(attributes).some(([, value]) => !!value);
                if (hasStyles) {
                    return true;
                }
                return commands.unsetMark(this.name);
            },
        };
    },
});

exports.TextStyle = TextStyle;
exports["default"] = TextStyle;
