// @tiptap/extension-table-header/dist/tiptap-extension-table-header.esm.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tiptapCore_esm = require('../../core/dist/tiptap-core.esm.js');

const TableHeader = tiptapCore_esm.Node.create({
    name: 'tableHeader',
    addOptions() {
        return {
            HTMLAttributes: {},
        };
    },
    content: 'block+',
    addAttributes() {
        return {
            colspan: {
                default: 1,
            },
            rowspan: {
                default: 1,
            },
            colwidth: {
                default: null,
                parseHTML: element => {
                    const colwidth = element.getAttribute('colwidth');
                    const value = colwidth
                        ? [parseInt(colwidth, 10)]
                        : null;
                    return value;
                },
            },
        };
    },
    tableRole: 'header_cell',
    isolating: true,
    parseHTML() {
        return [
            { tag: 'th' },
        ];
    },
    renderHTML({ HTMLAttributes }) {
        return ['th', tiptapCore_esm.mergeAttributes(this.options.HTMLAttributes, HTMLAttributes), 0];
    },
});

exports.TableHeader = TableHeader;
exports["default"] = TableHeader;
